fe/fe-core/src/main/java/com/starrocks/lake/qe/scheduler/DefaultSharedDataWorkerProvider.java

```
@@ -245,6 +247,29 @@ public boolean allowUsingBackupNode() {
         return true;
     }
 
+    // Makes best effort attempt to use the internal tablet mapper to select a backup based on tabletId
+    private Optional<Long> useInternalTabletMapperToSelectBackup(long workerId, Long tabletId) {
+        SystemInfoService systemInfoService = GlobalStateMgr.getCurrentState().getNodeMgr().getClusterInfo();
+        TabletComputeNodeMapper mapper = systemInfoService.internalTabletMapper();
+        List<Long> cnIdsOrderedByPreference =
+                mapper.backupComputeNodesForTablet(tabletId, workerId, availableID2ComputeNode.size()-1);
+        if (cnIdsOrderedByPreference == null) {
+            LOG.warn(String.format("The internal tablet mapper doesn't seem to know about the resource" +
+                            " isolation group %s. Its state is %s.",
+                    GlobalStateMgr.getCurrentState().getNodeMgr().getMySelf().getResourceIsolationGroup(), mapper.debugString()));
+            return Optional.empty();
+        }
+        for (Long possibleBackup : cnIdsOrderedByPreference) {
+            if (possibleBackup != workerId && availableID2ComputeNode.containsKey(possibleBackup)) {
+                return Optional.of(possibleBackup);
+            }
+        }
+        LOG.warn("Tried to use internal tablet to CN mapper but it failed to find an available cn for the given" +
```

If for whatever reason we aren't able to find a suitable CN, how would we know? Are we relying on tablets per node to make judgements about whether replica allocation is succeeding?

---

Fair to ask, this warning might go unnoticed and it's another design option to fail loudly here. CN_TO_OWNED_TABLET_COUNT wouldn't help us here. 

This function in particular is handling when a CN which ordinarily owns some tablet is unavailable. I choose to log the warning and fallback to using some other CN for backup since that seems preferable to failing loudly by throwing an exception, since that would cause errors for nearly all queries.

---

Btw, if we don't find any backup CN by falling back to the non-TabletComputeNodeMapper strategy, that will fail very loudly.

---

Makes sense. Agree on the reasoning for not failing loudly here.